app/org/sagebionetworks/bridge/services/TaskService.java

```diff
@@ -82,33 +82,34 @@ public void setSurveyResponseService(SurveyResponseService surveyResponseService
         this.surveyResponseService = surveyResponseService;
     }
     
-    public List<Task> getTasks(User user, DateTime endsOn) {
+    public List<Task> getTasks(User user, ScheduleContext context) {
         checkNotNull(user);
-        checkNotNull(endsOn);
+        checkNotNull(context);
         
-        DateTime now = DateTime.now();
-        if (endsOn.isBefore(now)) {
-            throw new BadRequestException("End timestamp must be after the time of the request");
-        } else if (endsOn.minusDays(MAX_EXPIRES_ON_DAYS).isAfter(now)) {
-            throw new BadRequestException("Task request window must be "+MAX_EXPIRES_ON_DAYS+" days or less");
-        }
+        Validate.nonEntityThrowingException(VALIDATOR, context);
+        
+        Map<String, DateTime> events = createEventsMap(context);
+        
+        // Get tasks from the scheduler. None of these tasks have been saved, some may be new,
+        // and some may have already been persisted. They are identified by their runKey.
+        List<Task> scheduledTasks = scheduleTasksForPlans(user, context.withEvents(events));
         
-        StudyIdentifier studyId = new StudyIdentifierImpl(user.getStudyKey());
         List<Task> tasksToSave = Lists.newArrayList();
-
-        Map<String,List<Task>> scheduledTasks = scheduleTasksForPlans(user, endsOn);
-        for (String runKey : scheduledTasks.keySet()) {
-            if (taskDao.taskRunHasNotOccurred(user.getHealthCode(), runKey)) {
-                for (Task task : scheduledTasks.get(runKey)) {
-                    Activity activity = createResponseActivityIfNeeded(
-                        studyId, user.getHealthCode(), task.getActivity());
-                    task.setActivity(activity);
-                    tasksToSave.add(task);
-                }
+        for (Task task : scheduledTasks) {
+            if (taskDao.taskRunHasNotOccurred(context.getHealthCode(), task.getRunKey())) {
```

Comment:
I'll restore the map, but they're 1:1 unless you have multiple activities in a schedule plan... so far no one has done that. All the activities will have the same key (as once you've calculated and saved one of them, you'll have calculated and saved all of them).